compiler/test/dotty/tools/dotc/reporting/TestReporter.scala

```diff
@@ -144,4 +152,15 @@ object TestReporter {
     }
     rep
   }
+
+  def lastRunFailedTests: List[String] =
```

Comment:
It looks like we make no distinction between the situation when `--failed` is not enabled and when this option is on but all the tests passed in the previous run. I think we should split these cases, e.g. by returning `None` in the former case and `Some(Nil)` in the latter. In general, if all the tests in the previous run passed, I would expect `testCompilation --failed` to run no tests and give explicit feedback that no tests were run instead of running all the tests.